// @kbn/securitysolution-rules/target_node/rule_type_constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.THRESHOLD_RULE_TYPE_ID = exports.SIGNALS_ID = exports.SAVED_QUERY_RULE_TYPE_ID = exports.QUERY_RULE_TYPE_ID = exports.ML_RULE_TYPE_ID = exports.INDICATOR_RULE_TYPE_ID = exports.EQL_RULE_TYPE_ID = void 0;

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/**
 * Id for the legacy siem signals alerting type
 */
const SIGNALS_ID = `siem.signals`;
/**
 * IDs for alerts-as-data rule types
 */

exports.SIGNALS_ID = SIGNALS_ID;
const RULE_TYPE_PREFIX = `siem`;
const EQL_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.eqlRule`;
exports.EQL_RULE_TYPE_ID = EQL_RULE_TYPE_ID;
const INDICATOR_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.indicatorRule`;
exports.INDICATOR_RULE_TYPE_ID = INDICATOR_RULE_TYPE_ID;
const ML_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.mlRule`;
exports.ML_RULE_TYPE_ID = ML_RULE_TYPE_ID;
const QUERY_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.queryRule`;
exports.QUERY_RULE_TYPE_ID = QUERY_RULE_TYPE_ID;
const SAVED_QUERY_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.savedQueryRule`;
exports.SAVED_QUERY_RULE_TYPE_ID = SAVED_QUERY_RULE_TYPE_ID;
const THRESHOLD_RULE_TYPE_ID = `${RULE_TYPE_PREFIX}.thresholdRule`;
exports.THRESHOLD_RULE_TYPE_ID = THRESHOLD_RULE_TYPE_ID;
