// @elastic/react-search-ui/lib/containers/Result.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ResultContainer = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = require("react");

var _reactSearchUiViews = require("@elastic/react-search-ui-views");

var _ = require("..");

var _types = require("../types");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ResultContainer = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ResultContainer, _Component);

  var _super = _createSuper(ResultContainer);

  function ResultContainer() {
    var _this;

    (0, _classCallCheck2.default)(this, ResultContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClickLink", function (id) {
      var _this$props = _this.props,
          clickThroughTags = _this$props.clickThroughTags,
          shouldTrackClickThrough = _this$props.shouldTrackClickThrough,
          trackClickThrough = _this$props.trackClickThrough;

      if (shouldTrackClickThrough) {
        trackClickThrough(id, clickThroughTags);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(ResultContainer, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          className = _this$props2.className,
          result = _this$props2.result,
          titleField = _this$props2.titleField,
          urlField = _this$props2.urlField,
          thumbnailField = _this$props2.thumbnailField,
          view = _this$props2.view,
          trackClickThrough = _this$props2.trackClickThrough,
          shouldTrackClickThrough = _this$props2.shouldTrackClickThrough,
          clickThroughTags = _this$props2.clickThroughTags,
          rest = (0, _objectWithoutProperties2.default)(_this$props2, ["className", "result", "titleField", "urlField", "thumbnailField", "view", "trackClickThrough", "shouldTrackClickThrough", "clickThroughTags"]);
      var View = view || _reactSearchUiViews.Result;
      return View(_objectSpread({
        className: className,
        result: result,
        key: "result-".concat(result.id.raw),
        onClickLink: function onClickLink() {
          return _this2.handleClickLink(result.id.raw);
        },
        titleField: titleField,
        urlField: urlField,
        thumbnailField: thumbnailField
      }, rest));
    }
  }]);
  return ResultContainer;
}(_react.Component);

exports.ResultContainer = ResultContainer;
(0, _defineProperty2.default)(ResultContainer, "propTypes", {
  // Props
  className: _propTypes.default.string,
  clickThroughTags: _propTypes.default.arrayOf(_propTypes.default.string),
  titleField: _propTypes.default.string,
  urlField: _propTypes.default.string,
  thumbnailField: _propTypes.default.string,
  view: _propTypes.default.func,
  result: _types.Result.isRequired,
  shouldTrackClickThrough: _propTypes.default.bool,
  // Actions
  trackClickThrough: _propTypes.default.func
});
(0, _defineProperty2.default)(ResultContainer, "defaultProps", {
  clickThroughTags: [],
  shouldTrackClickThrough: true
});

var _default = (0, _.withSearch)(function (_ref) {
  var trackClickThrough = _ref.trackClickThrough;
  return {
    trackClickThrough: trackClickThrough
  };
})(ResultContainer);

exports.default = _default;
