// @kbn/securitysolution-io-ts-alerting-types/target_node/threat_technique/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.threat_techniques = exports.threat_technique_reference = exports.threat_technique_name = exports.threat_technique_id = exports.threat_technique = void 0;

var t = _interopRequireWildcard(require("io-ts"));

var _threat_subtechnique = require("../threat_subtechnique");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/* eslint-disable @typescript-eslint/naming-convention */
const threat_technique_id = t.string;
exports.threat_technique_id = threat_technique_id;
const threat_technique_name = t.string;
exports.threat_technique_name = threat_technique_name;
const threat_technique_reference = t.string;
exports.threat_technique_reference = threat_technique_reference;
const threat_technique = t.intersection([t.exact(t.type({
  id: threat_technique_id,
  name: threat_technique_name,
  reference: threat_technique_reference
})), t.exact(t.partial({
  subtechnique: _threat_subtechnique.threat_subtechniques
}))]);
exports.threat_technique = threat_technique;
const threat_techniques = t.array(threat_technique);
exports.threat_techniques = threat_techniques;
