// @kbn/securitysolution-io-ts-list-types/target_web/common/list_operator/index.js
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import * as t from 'io-ts';
export var listOperator = t.keyof({
  excluded: null,
  included: null
});
export var ListOperatorEnum;

(function (ListOperatorEnum) {
  ListOperatorEnum["INCLUDED"] = "included";
  ListOperatorEnum["EXCLUDED"] = "excluded";
})(ListOperatorEnum || (ListOperatorEnum = {}));

export var ListOperatorTypeEnum;

(function (ListOperatorTypeEnum) {
  ListOperatorTypeEnum["NESTED"] = "nested";
  ListOperatorTypeEnum["MATCH"] = "match";
  ListOperatorTypeEnum["MATCH_ANY"] = "match_any";
  ListOperatorTypeEnum["WILDCARD"] = "wildcard";
  ListOperatorTypeEnum["EXISTS"] = "exists";
  ListOperatorTypeEnum["LIST"] = "list";
})(ListOperatorTypeEnum || (ListOperatorTypeEnum = {}));
