// @kbn/config/target_node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ConfigService", {
  enumerable: true,
  get: function () {
    return _config_service.ConfigService;
  }
});
Object.defineProperty(exports, "Env", {
  enumerable: true,
  get: function () {
    return _env.Env;
  }
});
Object.defineProperty(exports, "ObjectToConfigAdapter", {
  enumerable: true,
  get: function () {
    return _object_to_config_adapter.ObjectToConfigAdapter;
  }
});
Object.defineProperty(exports, "RawConfigService", {
  enumerable: true,
  get: function () {
    return _raw.RawConfigService;
  }
});
Object.defineProperty(exports, "applyDeprecations", {
  enumerable: true,
  get: function () {
    return _deprecation.applyDeprecations;
  }
});
Object.defineProperty(exports, "configDeprecationFactory", {
  enumerable: true,
  get: function () {
    return _deprecation.configDeprecationFactory;
  }
});
Object.defineProperty(exports, "getConfigFromFiles", {
  enumerable: true,
  get: function () {
    return _raw.getConfigFromFiles;
  }
});
Object.defineProperty(exports, "getPluginSearchPaths", {
  enumerable: true,
  get: function () {
    return _plugins.getPluginSearchPaths;
  }
});
Object.defineProperty(exports, "hasConfigPathIntersection", {
  enumerable: true,
  get: function () {
    return _config.hasConfigPathIntersection;
  }
});
Object.defineProperty(exports, "isConfigPath", {
  enumerable: true,
  get: function () {
    return _config.isConfigPath;
  }
});

var _deprecation = require("./deprecation");

var _raw = require("./raw");

var _config_service = require("./config_service");

var _config = require("./config");

var _object_to_config_adapter = require("./object_to_config_adapter");

var _env = require("./env");

var _plugins = require("./plugins");
